File: symbol_ir.py
from span import SourceSpan
from ir_node import IRNode
from expression_ir import IdentifiedIRNode


class SymbolIR(IdentifiedIRNode):
    def __init__(self, id: int, name: str, kind: str, scope_id: int, span: SourceSpan):
        super().__init__(id=id, span=span)
        self.id = id
        self.name = name
        self.kind = kind
        self.scope_id = scope_id
        self.span = span

File: rust/src/ir/nodes/symbol_ir.rs
use crate::ir::{span_ir::SourceSpan};

#[derive(Debug, Clone)]
pub struct SymbolIR {
    pub id: i64,
    pub name: String,
    pub kind: SymbolKind,
    pub scope_id: i64,
    pub span: Option<SourceSpan>,
}

#[derive(Debug, Clone)]
pub enum SymbolKind {
    Unknown,
    Variable,
    Function,
    Class,
    Param,
    ModuleAlias
}

impl From<i32> for SymbolKind {
    fn from(value: i32) -> Self {
        match value {
            1 => Self::Variable,
            2 => Self::Function,
            3 => Self::Class,
            4 => Self::Param,
            5 => Self::ModuleAlias,
            _ => Self::Unknown,
        }
    }
}
